### ISSUE TITLE
Attempting to move twice in a row

### ISSUE THREAD

**Issue by roy7** [online-go/gtp2ogs#18]
My logs aren't the best for this... but basically K6 was a valid move. However the bot is asked to make another move before the opponent had gone. It came up with N7, which turned out to the move the opponent already made. Since the move isn't valid the server ignores it, but gtp2ogs just sits there waiting. I log most events, but nothing seemed to arrive before the makeMove 46 happened. Is it possible the SetInterval "when bored" code could cause a spurious turn to be taken when it isn't the bot's move?

  gtp2ogs.js:209                 [14972] <<< = K6
  gtp2ogs.js:688                 [Game 1911] Playing k6 { x: 9, y: 13, text: 'k6', resign: false, pass: false }
  gtp2ogs.js:1174                #  clearTimeout 1911
  gtp2ogs.js:1174                #  setTimeout 1911
  gtp2ogs.js:688                 [Game 1911] clock
  gtp2ogs.js:209                 [14972] >>> kgs-time_settings byoyomi 600 29 5
  gtp2ogs.js:209                 [14972] >>> time_left black 444 0
  gtp2ogs.js:209                 [14972] >>> time_left white 207 0
  gtp2ogs.js:688                 [Game 1911] clock
  gtp2ogs.js:209                 [14972] >>> kgs-time_settings byoyomi 600 29 5
  gtp2ogs.js:209                 [14972] >>> time_left black 435 0
  gtp2ogs.js:209                 [14972] >>> time_left white 207 0
  gtp2ogs.js:1174                #  Connected to game 1911 already
  gtp2ogs.js:688                 [Game 1911] makeMove 46
  gtp2ogs.js:209                 [14972] Generating move for game 1911
  gtp2ogs.js:209                 [14972] >>> genmove white
  gtp2ogs.js:1174                #  clearTimeout 1911
  gtp2ogs.js:1174                #  setTimeout 1911
  gtp2ogs.js:1174                #  clearTimeout 1911
  gtp2ogs.js:1174                #  setTimeout 1911
  gtp2ogs.js:209                 [14972] <<< = N7
  gtp2ogs.js:688                 [Game 1911] Playing n7 { x: 12, y: 12, text: 'n7', resign: false, pass: false }
`

The actual game played was

https://beta.online-go.com/game/1911

**Comment by roy7** [comment]
Extra debugging data shows me when it happened again, sendMove() wasn't called for an opponent's move. The move event came in, but sendMove() not called. So move_were_waiting_for must have been wrong somehow. Then when the next active_game comes in, it is our turn and we move... but the bot never saw the prior opponant move and thinks that space is blank rest of the game until an illegal turn much later when it tried to go there.

https://beta.online-go.com/game/1947

Move 197 to R12 was invisible to the bot. I will start logging move_number_were_waiting_for each time a move event arrives.

**Comment by roy7** [comment]
Breakthrough... active_game always arrives before move event... except when it breaks. That almost never happens. Looking over the log, active_game events keep arriving before move events. But the time it broke, the move event arrived before the active_game event.

**Comment by roy7** [comment]
this.waiting_on_gamedata_to_make_move isn't set until active_game arrives. If move is after active_game, then check_for_move makes the move. If you reverse it, check_for_move doesn't make the move.

Should we store "move we are waiting for" after we move (since we know our move number we just made), and change

`if (this.waiting_on_gamedata_to_make_move && this.state.moves.length == this.move_number_were_waiting_for)`

to something like

`if (this.state.moves.length == this.move_number_were_waiting_for)`

And remove the check for waiting on gamedata? If we know the move we need, and that move has arrived (length = waiting for) then send it to the bot?

**Comment by roy7** [comment]
Ok I think this is a bug of my own invention in the --persist branch. When a move arrives in the current normal --devel tree, it is sent to the bot. Because when you start over every time, you either need to move or wait. If you need to wait, the next move to arrive goes to bot. Once you move, bot is killed.

To persist, I had to check for moves to pass along in that area but not send my own moves. I did this:

`if (this.move_number_were_waiting_for == move.move_number)`

naively thinking this.move_number_were_waiting_for was a global I could use safely for this purpose. But it isn't if a move arrives before active_game, which can't happen in a "restart every move" world where you get full gamedata events instead.

The solution may just be a better check here for when to send moves to the bot, not using this.move_number_were_waiting_for.

**Comment by roy7** [comment]
Anoek, a solution for my --persist branch that I tested some and seems straight forward is to change

`this.move_number_were_waiting_for += 2;`

to 

`this.move_number_were_waiting_for += 2;`

That seems to work for each case I tried it with and shouldn't break any existing code that I can tell. And when we do get the move we need from opponent, we know the next one we need will always be 2 moves away. I think this should solve the active_game vs move event race condition. Thoughts?
